### PR TITLE
fix(appstore): Ensure returned apps from AppStore are valid

### LINKED ISSUE #47844: [Bug]: Upgrade from 29.0.4 to 29.0.6 fails with TypeError in AppFetcher.php
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

```
Should the "occ upgrade" command be executed? [Y/n] y
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Setting log level to debug
Repair step: Repair MySQL collation
Repair info: All tables already have the correct collation -> nothing to do
Repair step: Copy data from accounts table when migrating from ownCloud
Repair step: Drop account terms table when migrating from ownCloud
Updating database schema
Updated database
An unhandled exception has been thrown:
TypeError: array_filter(): Argument #1 ($array) must be of type array, null given in /var/www/html/lib/private/App/AppStore/Fetcher/AppFetcher.php:160
Stack trace:
#0 /var/www/html/lib/private/App/AppStore/Fetcher/AppFetcher.php(160): array_filter(NULL)
#1 /var/www/html/lib/private/App/AppStore/Fetcher/Fetcher.php(182): OC\App\AppStore\Fetcher\AppFetcher->fetch('', '', false)
#2 /var/www/html/lib/private/App/AppStore/Fetcher/AppFetcher.php(179): OC\App\AppStore\Fetcher\Fetcher->get(false)
#3 /var/www/html/lib/private/Installer.php(405): OC\App\AppStore\Fetcher\AppFetcher->get(false)
#4 /var/www/html/lib/private/Updater.php(415): OC\Installer->isUpdateAvailable('calendar')
#5 /var/www/html/lib/private/Updater.php(275): OC\Updater->upgradeAppStoreApps(Array)
#6 /var/www/html/lib/private/Updater.php(133): OC\Updater->doUpgrade('29.0.6.1', '29.0.4.1')
#7 /var/www/html/core/Command/Upgrade.php(216): OC\Updater->upgrade()
#8 /var/www/html/3rdparty/symfony/console/Command/Command.php(298): OC\Core\Command\Upgrade->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#9 /var/www/html/3rdparty/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#10 /var/www/html/3rdparty/symfony/console/Application.php(301): Symfony\Component\Console\Application->doRunCommand(Object(OC\Core\Command\Upgrade), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#11 /var/www/html/3rdparty/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#12 /var/www/html/lib/private/Console/Application.php(213): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 /var/www/html/console.php(102): OC\Console\Application->run()
#14 /var/www/html/occ(11): require_once('/var/www/html/c...')
#15 {main}
```

Looking at the source code, this happens because the app store server is not returning the expected data. Are there issues with the app store currently?

### Steps to reproduce

Run occ upgrade

### Expected behavior

N/A

### Nextcloud Server version

29

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by solracsf**
* Resolves: https://github.com/nextcloud/server/issues/47844

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by come-nc — COMMENTED** [review]
This does not fix the linked TypeError from what I understand.
The problem of the TypeError is more that `$response['data']` was not set I suppose?
Any idea what the response looked like? Because if it was empty it would return early line 57 already.

**Comment by provokateurin** [comment]
No @come-nc's analysis is right:
```
<?php
array_filter(null);
```
will give you
```
PHP Fatal error:  Uncaught TypeError: array_filter(): Argument #1 ($array) must be of type array, null given in /home/jld3103/.config/JetBrains/IntelliJIdea2024.2/scratches/scratch_4.php:2
Stack trace:
#0 /home/jld3103/.config/JetBrains/IntelliJIdea2024.2/scratches/scratch_4.php(2): array_filter()
#1 {main}
  thrown in /home/jld3103/.config/JetBrains/IntelliJIdea2024.2/scratches/scratch_4.php on line 2

Process finished with exit code 255
```
Which is exactly the same error message. The error message clearly mentions the method and the argument.

**Review by provokateurin — CHANGES_REQUESTED** [review]
The `if (empty($response)) {` check further up should be extended to be `if (empty($response) || $response['data'] === null) {`.

**Comment by provokateurin** [comment]
> However, there may be cases where $response['data'] elements are null or invalid, causing array_filter() to throw an error.

If no callback is given to array_filter it will remove all falsy values which includes null, so null is not an invalid element in the given array.

**Comment by solracsf** [comment]
Thanks for your inputs, six eyes see better than two 🤣   
PR amended.

**Review by provokateurin — APPROVED** [review]

**Comment by provokateurin** [comment]
I wonder if we should have a warning log for the return case to inform the admin that the appstore returned a faulty response? Otherwise it is just silent and they can't know why.

**Comment by solracsf** [comment]
Agree.

Maybe something like:

```
$this->logger->warning('Response from appstore is invalid, apps could not be retrieved. Try again later.', ['app' => 'appstoreFetcher']);
```

This should inform admin that response was not valid, but also this could be temporary and it can be retried (in case appstore is down for maintenance, overloaded...).

**Comment by provokateurin** [comment]
Sounds good, but I would either omit the app parameter or set it to core.

**Comment by solracsf** [comment]
All the other warnings have it, that's why I've included it. 🤔

**Comment by provokateurin** [comment]
Ah ok, then it's good. Can you add it to the PR?

**Comment by solracsf** [comment]
Done ✅

File: lib/private/App/AppStore/Fetcher/AppFetcher.php [hunk 1]
```diff
@@ -53,7 +53,8 @@ protected function fetch($ETag, $content, $allowUnstable = false) {
 		/** @var mixed[] $response */
 		$response = parent::fetch($ETag, $content);
 
-		if (empty($response)) {
+		if (empty($response) || $response['data'] === null) {
```
[the remark this answers is not in the document]
Comment (come-nc) [reply lib/private/App/AppStore/Fetcher/AppFetcher.php:56]: empty and isset both emit no notice for missing key/var.
But empty is ugly, let’s go with first version.

**Review by come-nc — APPROVED** [review]

**Pull request merged**